api/handlers.go

```diff
@@ -165,3 +171,404 @@ func (api *RecipeAPI) AddRecipeHandler(w http.ResponseWriter, req *http.Request)
 	w.Header().Set("content-type", "application/json")
 	w.Write(output)
 }
+
+//AddInstanceHandler - add an instance to an existing recipe
+func (api *RecipeAPI) AddInstanceHandler(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+
+	//Get Update Recipe ID
+	vars := mux.Vars(req)
+	recipeID := vars["id"]
+	logD := log.Data{"recipe_id": recipeID}
+
+	// Read body
+	b, err := ioutil.ReadAll(req.Body)
+	defer req.Body.Close()
+	if err != nil {
+		log.Event(ctx, "error in reading request body", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Unmarshal to the shape of Instance struct
+	var instance recipe.Instance
+	err = json.Unmarshal(b, &instance)
+	if err != nil {
+		log.Event(ctx, "error returned from json unmarshal", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	//Validation to check if all the instance fields are entered
+	err = instance.ValidateAddInstance()
+	if err != nil {
+		log.Event(ctx, "bad request error as invalid instance given in request body", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	// Get currentRecipe from ID given to get instance of the recipe
+	currentRecipe, err := api.dataStore.Backend.GetRecipe(recipeID)
+	if err != nil {
+		log.Event(ctx, "error retrieving specific recipe from mongo", log.ERROR, log.Error(err), logD)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Append new instance into the current instance of the recipe
+	currentRecipe.OutputInstances = append(currentRecipe.OutputInstances, instance)
+
+	// Update the current recipe in mongo with the updated instance
+	update := bson.M{"$set": currentRecipe}
```

Comment:
See longer comment on mongo package encapsulation

Reply:
Done in commit "move bson statements to mongo package - mongo package encapsulation"